### PR TITLE
feature(new integration): Canny Source Onboarding

### PR DESCRIPTION
**Pull request by rohithkaza**
## Description of the change

> Onboarding Canny Source

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: v0/sources/canny/mapping.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[
+  {
+    "sourceKeys": "created",
+    "destKeys": "timestamp"
```
Comment (ItsSudip) [inline v0/sources/canny/mapping.json:4]: Add `originalTimestamp` aswell.

File: v0/sources/canny/transform.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  }
+
+  if (event.cannyId) {
+    message.context.externalId = [
+      {
+        type: "cannyId",
+        value: event.cannyId
+      }
+    ];
+  }
```
Comment (ItsSudip) [inline v0/sources/canny/transform.js:47]: Can you explain what is the purpose of this? As canny doesn't send anything named cannyId.

File: v0/sources/canny/transform.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      });
+    } else {
+      message.context.externalId = [
+        {
+          type: "authorId",
+          value: event.object?.author?.id || event.object?.post?.author?.id
+        }
+      ];
+    }
+  }
```
Comment (ItsSudip) [inline v0/sources/canny/transform.js:62]: We can move this to the previous section. As we have to choose from object.voter/object.author. The id should be available always.

File: v0/sources/canny/transform.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  if (event.object?.author?.id || event.object?.post?.author?.id) {
+    if (message.context.externalId) {
+      message.context.externalId = message.context.externalId.push({
+        type: "authorId",
+        value: event.object?.author?.id || event.object?.post?.author?.id
+      });
+    } else {
+      message.context.externalId = [
+        {
+          type: "authorId",
```
Comment (ItsSudip) [inline v0/sources/canny/transform.js:57]: And make this to `cannyId`.

File: v0/sources/canny/mapping.json (hunk 1 above)
Comment (rohithkaza) [reply v0/sources/canny/mapping.json]: Done

File: v0/sources/canny/transform.js (hunk 2 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: It's a mistake. Removed it

File: v0/sources/canny/transform.js (hunk 3 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/transform.js (hunk 4 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/transform.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+// import mapping json using JSON.parse to preserve object key order
+const mapping = JSON.parse(
+  fs.readFileSync(path.resolve(__dirname, "./mapping.json"), "utf-8")
+);
+
+function process(event) {
+  const message = new Message(`Canny`);
+
+  // event type is always track
+  const eventType = "track";
```
Comment (ItsSudip) [inline v0/sources/canny/transform.js:15]: We can remove this and hardcode line 17 to "track"

File: v0/sources/canny/transform.js [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        message.context.traits = event.object?.author;
+
+        // deleting already mapped fields
+        delete message.properties?.author;
+        delete message.context.traits?.userID;
+        delete message.context.traits?.id;
+    }
+  } catch (e) {
+    throw new Error(`Missing essential fields from Canny. Error: (${e})`);
+  }
```
Comment (ItsSudip) [inline v0/sources/canny/transform.js:78]: We can remove this switch with a function something like createMessage with parameters `event` and `typeOfUser` i.e. `voter` or `author`.

File: v0/sources/canny/mapping.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+[
+  {
+    "sourceKeys": "created",
+    "destKeys": ["timestamp", "originalTimestamp"]
+  },
+  {
+    "sourceKeys": "object",
+    "destKeys": "properties"
```
Comment (utsabc) [inline v0/sources/canny/mapping.json:8]: You can exclude `author` or `voter` when mapping things from `object` 
example: https://github.com/rudderlabs/rudder-transformer/blob/a8560035932b96a18da0e37b98d3c25f149739ac/v0/destinations/vero/data/VeroIdentifyConfig.json#L16

File: v0/sources/canny/transform.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+  message.setEventType(eventType);
+
+  message.setPropertiesV2(event, mapping);
+
+  message.context.integration.version = "1.0.0";
+  try {
+    switch (event.type) {
+      case "vote.created":
+      case "vote.deleted":
```
Comment (utsabc) [inline v0/sources/canny/transform.js:25]: For these we can use separate mapping things will be cleaner in terms of code

File: v0/sources/canny/transform.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            }
+          ];
+        }
+
+        message.context.traits = event.object?.voter;
+
+        // deleting already mapped fields
+        delete message.properties?.voter;
+        delete message.context.traits?.userID;
+        delete message.context.traits?.id;
```
Comment (utsabc) [inline v0/sources/canny/transform.js:48]: This can be handled during mapping using excludes field?

File: v0/sources/canny/transform.js [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        break;
+
+      default:
+        // setting up userId
```
Comment (utsabc) [inline v0/sources/canny/transform.js:52]: Again for `author` a separate mapping can be used

**Review by utsabc — CHANGES_REQUESTED** [review]

File: v0/sources/canny/transform.js (hunk 5 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/transform.js (hunk 6 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js:42]: Done

File: v0/sources/canny/transform.js [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+  try {
+    switch (event.type) {
+      case "vote.created":
+      case "vote.deleted":
+        // setting up userId
+        if (event.object?.voter?.userID) {
+          message.userId = event.object.voter.userID;
+        } else {
+          // setting up anonymousId if userId is not present
+          message.anonymousId = sha256(event.object.voter?.email);
```
[the remark this answers is not in the document]
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: catching the error at line 41

File: v0/sources/canny/transform.js (hunk 8 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Added

File: v0/sources/canny/transform.js (hunk 10 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/transform.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+ * @param {*} typeOfUser
+ * @returns message
+ */
+function createMessage(event, typeOfUser) {
+  const message = new Message(`Canny`);
+
+  message.setEventType("track");
+
+  if (typeOfUser === "voter") message.setPropertiesV2(event, voterMapping);
+  else message.setPropertiesV2(event, authorMapping);
```
Comment (ItsSudip) [inline v0/sources/canny/transform.js:57]:
```suggestion
  if (typeOfUser === "voter"){ 
  message.setPropertiesV2(event, voterMapping);
  }
  else {
  message.setPropertiesV2(event, authorMapping);
  }
```

File: v0/sources/canny/voterMapping.json [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+  {
+    "sourceKeys": "created",
+    "destKeys": ["timestamp", "originalTimestamp"]
+  },
+  {
+    "sourceKeys": "object",
+    "destKeys": "properties",
+    "metadata": {
+      "excludes": ["voter"]
+    }
```
Comment (ItsSudip) [inline v0/sources/canny/voterMapping.json:11]: We are not using metadata in `setPropertiesV2` function. So it is not needed.

File: v0/sources/canny/voterMapping.json [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+    "sourceKeys": "type",
+    "destKeys": "event",
+    "required": true
+  },
+  {
+    "sourceKeys": "object.voter",
+    "destKeys": "context.traits",
+    "metadata": {
+      "excludes": ["userID", "id"]
+    }
```
Comment (ItsSudip) [inline v0/sources/canny/voterMapping.json:27]: Same here

File: v0/sources/canny/transform.js (hunk 9 above)
Comment (ItsSudip) [reply v0/sources/canny/transform.js]: We don't have that functionality in `setPropertiesV2` function.

File: v0/sources/canny/mapping.json (hunk 7 above)
Comment (rohithkaza) [reply v0/sources/canny/mapping.json]: "excludes" is not working

File: v0/sources/canny/transform.js (hunk 9 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: excludes is not working

File: v0/sources/canny/transform.js (hunk 12 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/voterMapping.json (hunk 13 above)
Comment (rohithkaza) [reply v0/sources/canny/voterMapping.json]: Ok, removed it

File: v0/sources/canny/voterMapping.json (hunk 14 above)
Comment (rohithkaza) [reply v0/sources/canny/voterMapping.json]: Done

**Review by ItsSudip — DISMISSED** [review]

File: v0/sources/canny/transform.js [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+      typeOfUser = "voter";
+      break;
+
+    default:
+      typeOfUser = "author";
+  }
+
+  const message = createMessage(event, typeOfUser);
+
+  return message;
```
Comment (utsabc) [inline v0/sources/canny/transform.js:100]:
```suggestion
 return  createMessage(event, typeOfUser);

```

File: v0/sources/canny/transform.js [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+function process(event) {
+  let typeOfUser;
+
+  switch (event.type) {
+    case "vote.created":
+    case "vote.deleted":
```
Comment (utsabc) [inline v0/sources/canny/transform.js:90]:
```suggestion
    case CannyOperation.VOTE_CREATED:
    case CannyOperation.VOTE_DELETED:
```

Where
```
const CannyOperation = {
VOTE_CREATED: "vote.created",
VOTE_DELETED: "vote.deleted"
}
```

File: v0/sources/canny/transform.js [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+  try {
+    // setting up userId
+    if (event.object[`${typeOfUser}`]?.userID) {
+      message.userId = event.object[`${typeOfUser}`].userID;
+    } else {
+      // setting up anonymousId if userId is not present
+      message.anonymousId = sha256(event.object[`${typeOfUser}`]?.email);
+    }
+  } catch (e) {
+    throw new Error(`Missing essential fields from Canny. Error: (${e})`);
```
Comment (utsabc) [inline v0/sources/canny/transform.js:41]: Also put a log here

File: v0/sources/canny/transform.js [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+  checkForRequiredFields(message);
+
+  if (event.object[`${typeOfUser}`]?.id) {
+    message.context.externalId = [
+      {
+        type: "cannyUserId",
+        value: event.object[`${typeOfUser}`].id
+      }
+    ];
+  }
```
Comment (utsabc) [inline v0/sources/canny/transform.js:75]: Can we move this to `settingIds` fxn

**Review by utsabc — CHANGES_REQUESTED** [review]

File: v0/sources/canny/transform.js (hunk 15 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/transform.js (hunk 16 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Done

File: v0/sources/canny/transform.js (hunk 17 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js:41]: Done

File: v0/sources/canny/transform.js (hunk 18 above)
Comment (rohithkaza) [reply v0/sources/canny/transform.js]: Moved

**Review by utsabc — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**